internal/components/receivers/prometheus.go

```diff
@@ -0,0 +1,160 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package receivers
+
+import (
+	"github.com/go-logr/logr"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/open-telemetry/opentelemetry-operator/internal/naming"
+)
+
+type namespaces struct {
+	Names []string `mapstructure:"names"`
+}
+
+type kubernetesSDConfig struct {
+	Namespaces namespaces `mapstructure:"namespaces"`
+	Role       string     `mapstructure:"role"`
+}
+
+type scrapeConfig struct {
+	KubernetesSDConfigs []kubernetesSDConfig `mapstructure:"kubernetes_sd_configs"`
+	JobName             string               `mapstructure:"job_name"`
+}
+
+type prometheusConfig struct {
+	ScrapeConfigs *[]scrapeConfig `mapstructure:"scrape_configs"`
+}
+
+type prometheusReceiverConfig struct {
+	Config *prometheusConfig `mapstructure:"config"`
+}
+
+func generatePrometheusReceiverRoles(logger logr.Logger, config prometheusReceiverConfig, componentName string, otelCollectorName string) ([]*rbacv1.Role, error) {
+	if config.Config == nil {
+		return nil, nil
+	}
+
+	if config.Config.ScrapeConfigs == nil {
+		return nil, nil
+	}
```

Comment:
Should this cover the target allocator as well?

Reply:
For the TA we should do it in a different PR... otherwise this will become huge